### PR TITLE
ninja backend: Fix header deps of generated source compiles

### PR DESCRIPTION
**Pull request by nirbheek**
Because we were iterating over all generated sources of a target in one go, we weren't adding the correct `header_deps` for those. We need to aggregate the `header_deps` first before generating compile targets for generated source files.

Fixes https://github.com/mesonbuild/meson/issues/714

### CODE REVIEW

**Comment by nirbheek** [comment]
The test uses subprojects because that's how we encountered it in gstreamer, but it doesn't have to be for a subproject. An internal dependency within the same project will also trigger it.

**Comment by jpakkane** [comment]
Merge conflicts now. :(

**Comment by nirbheek** [comment]
Rebased!

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1553,6 +1572,8 @@ def get_link_debugfile_args(self, linker, target, outname):
         return linker.get_link_debugfile_args(outname)
 
     def generate_single_compile(self, target, outfile, src, is_generated=False, header_deps=[], order_deps=[]):
+        if 'gstrtsp-enumtypes.c' in str(src) or 'gstrtsptransport.c' in str(src):
+            print(src, header_deps, order_deps)
```
Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: Leftover debug code?

Comment (nirbheek) [inline mesonbuild/backend/ninjabackend.py]: Fixed! Sorry.

**Pull request merged**